python_modules/dagster/dagster/_core/definitions/partition_mapping.py

```diff
@@ -551,6 +555,309 @@ def get_downstream_partitions_for_partitions(
+                        dimension_mapping.partition_mapping.get_downstream_partitions_for_partitions(
+                            a_dimension_partitions_def.empty_subset().with_partition_keys([key]),
+                            b_dimension_partitions_def,
+                            # Add dynamic partitions store
+                        ).get_partition_keys()
+                    )
+        else:
+            # a_partitions_def is downstream of b_partitions_def, so we need to map the
+            # dimension names of a_partitions_def to the corresponding dependency dimensions of
+            # b_partitions_def
+            a_dim_to_dependency_b_dim = {
+                dimension_mapping.downstream_dimension_name: (
+                    upstream_dim,
+                    dimension_mapping,
+                )
+                for upstream_dim, dimension_mapping in self.downstream_mappings_by_upstream_dimension.items()
+            }
+
+            for dimension_name, keys in a_partition_keys_by_dimension.items():
+                (
+                    b_dim_name,
+                    dimension_mapping,
+                ) = a_dim_to_dependency_b_dim[dimension_name]
+                a_dimension_partitions_def = a_dimension_partitions_defs_by_name[dimension_name]
+                b_dimension_partitions_def = b_dimension_partitions_def_by_name[b_dim_name]
+                for key in keys:
+                    dep_b_keys_by_a_dim_and_key[dimension_name][key] = list(
+                        dimension_mapping.partition_mapping.get_upstream_partitions_for_partitions(
+                            a_dimension_partitions_def.empty_subset().with_partition_keys([key]),
+                            b_dimension_partitions_def,
+                            # Add dynamic partitions store
+                        ).get_partition_keys()
+                    )
+
+        b_partition_keys = set()
+        dimension_names = [dim.name for dim in a_partitions_def.partitions_defs]
+        for key in a_partition_keys:
+            key = cast(MultiPartitionKey, key)
+            for b_key_values in itertools.product(
+                *[
+                    dep_b_keys_by_a_dim_and_key[dim][key.keys_by_dimension[dim]]
+                    for dim in dimension_names
+                ]
+            ):
+                b_partition_keys.add(
+                    MultiPartitionKey(
+                        {dimension_names[i]: key for i, key in enumerate(b_key_values)}
+                    )
+                )
+
+        return b_partitions_def.empty_subset().with_partition_keys(b_partition_keys)
+
+    def get_upstream_partitions_for_partitions(
+        self,
+        downstream_partitions_subset: Optional[PartitionsSubset],
+        upstream_partitions_def: PartitionsDefinition,
+        dynamic_partitions_store: Optional[DynamicPartitionsStore] = None,
+    ) -> PartitionsSubset:
+        check.opt_inst_param(
```

Comment:
Because this is an internal method, I think we can rely on the type annotations instead of these checks